File: backend-go/internal/tools/dsl.go
package tools

import (
	"context"
	"encoding/json"
	"strconv"
)

type DSLCapabilityClient interface {
	GenerateDSL(
		ctx context.Context,
		actorUserID int64,
		runID string,
		projectID int64,
		conversationID string,
		arguments json.RawMessage,
	) (json.RawMessage, error)
}

type GenerateDSLTool struct {
	client DSLCapabilityClient
}

func NewGenerateDSLTool(client DSLCapabilityClient) GenerateDSLTool {
	return GenerateDSLTool{client: client}
}

func (t GenerateDSLTool) Definition() Definition {
	return Definition{
		Name: "generate_dsl",
		Description: "Validate and persist a research-v1 Action IR case authored from the user's goal and verified page elements. " +
			"The plan_binding must match the current persisted task plan. " +
			"Every step action must be one of: goto, click, input, wait_for, assert_text, " +
			"assert_url_contains, capture_text. Every step must declare intent, preconditions, postconditions, idempotency, and side_effect. " +
			"Author only semantic target and optional page_state/target_strategy fields; never author selector, candidates, or locator_confidence. " +
			"Locator preflight derives executable candidates from the submitted accessibility evidence. " +
			"Use preconditions for required pre-action state and postconditions for outcomes. " +
			"network_request conditions may match URL substring, method, and status on one observed event. " +
			"Express standalone visibility checks as wait_for, " +
			"and give every cross-page anchor click a url_contains postcondition with the expected destination URL or path. " +
			"never as assert_visible. Targets must be grounded in verified page elements. " +
			"When a goal requires search through page controls, preserve the verified input then click steps; " +
			"do not replace them with goto to a constructed search-result URL. " +
			"Input trigger is optional and only accepts Enter or Tab; omit it for ordinary semantic input, " +
			"and use a separate click step for a search button.",
		InputSchema: json.RawMessage(`{
			"type":"object",
			"additionalProperties":false,
			"properties":{
				"plan_binding":{
					"type":"object",
					"additionalProperties":false,
					"properties":{
						"plan_id":{"type":"string","minLength":1},
						"version":{"type":"integer","minimum":1},
						"sha256":{"type":"string","pattern":"^[0-9a-f]{64}$"}
					},
					"required":["plan_id","version","sha256"]
				},
				"case":{"$ref":"#/$defs/case"},
				"a11y_nodes_by_state":{
					"type":"object",
					"patternProperties":{"^.+$":{"type":"array","items":{"$ref":"#/$defs/a11y_node"}}},
					"additionalProperties":false
				}
			},
			"required":["plan_binding","case","a11y_nodes_by_state"],
			"$defs":{
				"condition":{
					"type":"object",
					"additionalProperties":false,
					"properties":{
						"type":{"type":"string","enum":["url_contains","url_changes","text_visible","text_gone","element_visible","element_gone","network_request","dom_changed","value_changed"]},
						"value":{"type":["string","null"]},
						"method":{"type":["string","null"],"minLength":1},
						"status":{"type":["integer","null"],"minimum":100,"maximum":599},
						"timeout_ms":{"type":"integer","minimum":100,"maximum":30000}
					},
					"required":["type"],
					"allOf":[
						{
							"if":{"properties":{"type":{"const":"url_contains"}},"required":["type"]},
							"then":{"properties":{"value":{"type":"string","minLength":1}},"required":["value"]}
						},
						{
							"if":{"properties":{"type":{"const":"network_request"}},"required":["type"]},
							"then":{"anyOf":[
								{"properties":{"value":{"type":"string","minLength":1}},"required":["value"]},
								{"properties":{"method":{"type":"string","minLength":1}},"required":["method"]},
								{"properties":{"status":{"type":"integer","minimum":100,"maximum":599}},"required":["status"]}
							]}
						}
					]
				},
				"input_contract":{
					"type":"object",
					"additionalProperties":false,
					"properties":{
						"name":{"type":"string","minLength":1,"maxLength":100},
						"context_key":{"type":"string","minLength":1,"maxLength":100,"pattern":"^[A-Za-z_][A-Za-z0-9_]*$"},
						"value_type":{"type":"string","enum":["string","number","boolean","object","array"]},
						"description":{"type":["string","null"]},
						"required":{"type":"boolean"},
						"value":{"type":["string","null"]}
					},
					"required":["name","context_key","value_type"]
				},
				"output_contract":{
					"type":"object",
					"additionalProperties":false,
					"properties":{
						"name":{"type":"string","minLength":1,"maxLength":100},
						"context_key":{"type":"string","minLength":1,"maxLength":100,"pattern":"^[A-Za-z_][A-Za-z0-9_]*$"},
						"value_type":{"type":"string","enum":["string","number","boolean","object","array"]},
						"description":{"type":["string","null"]},
						"source":{"type":["string","null"],"enum":["latest_url","error_message","status","last_step_url","last_step_page_title","last_step_target","last_step_value","last_step_error_message",null]}
					},
					"required":["name","context_key","value_type"]
				},
				"case":{
					"type":"object",
					"additionalProperties":false,
					"properties":{
						"profile":{"type":"string","const":"research-v1","default":"research-v1"},
						"name":{"type":"string","minLength":1,"maxLength":200},
						"description":{"type":["string","null"],"minLength":1,"maxLength":1000},
						"base_url":{"type":["string","null"],"minLength":1,"maxLength":500},
						"input_contract":{"type":"array","items":{"$ref":"#/$defs/input_contract"}},
						"output_contract":{"type":"array","items":{"$ref":"#/$defs/output_contract"}},
						"steps":{"type":"array","minItems":1,"items":{"oneOf":[
							{"$ref":"#/$defs/goto_step"},
							{"$ref":"#/$defs/click_step"},
							{"$ref":"#/$defs/input_step"},
							{"$ref":"#/$defs/wait_for_step"},
							{"$ref":"#/$defs/assert_text_step"},
							{"$ref":"#/$defs/assert_url_step"},
							{"$ref":"#/$defs/capture_text_step"}
						]}}
					},
					"required":["profile","name","steps"]
				},
				"goto_step":{
					"type":"object","additionalProperties":false,
					"properties":{
						"action":{"const":"goto"},"intent":{"type":"string","minLength":1,"maxLength":500},
						"target":{"type":"string","minLength":1},"value":{"type":"string","minLength":1},
						"preconditions":{"type":"array","items":{"$ref":"#/$defs/condition"}},
						"postconditions":{"type":"array","minItems":1,"items":{"$ref":"#/$defs/condition"}},
						"idempotency":{"const":"idempotent"},"side_effect":{"const":"browser_state"}
					},
					"required":["action","intent","target","value","preconditions","postconditions","idempotency","side_effect"]
				},
				"click_step":{
					"type":"object","additionalProperties":false,
					"properties":{
						"action":{"const":"click"},"intent":{"type":"string","minLength":1,"maxLength":500},
						"target":{"type":"string","minLength":1},"page_state":{"type":["string","null"]},
						"target_strategy":{"$ref":"#/$defs/target_strategy"},
						"preconditions":{"type":"array","minItems":1,"items":{"$ref":"#/$defs/condition"}},
						"postconditions":{"type":"array","minItems":1,"items":{"$ref":"#/$defs/condition"}},
						"idempotency":{"type":"string","enum":["idempotent","non_idempotent"]},
						"side_effect":{"type":"string","enum":["none","browser_state","external_state","unknown"]}
					},
					"required":["action","intent","target","preconditions","postconditions","idempotency","side_effect"]
				},
				"input_step":{
					"type":"object","additionalProperties":false,
					"properties":{
						"action":{"const":"input"},"intent":{"type":"string","minLength":1,"maxLength":500},
						"target":{"type":"string","minLength":1},"value":{"type":"string"},
						"trigger":{"type":["string","null"],"enum":["Enter","Tab",null]},
						"page_state":{"type":["string","null"]},"target_strategy":{"$ref":"#/$defs/target_strategy"},
						"preconditions":{"type":"array","minItems":1,"items":{"$ref":"#/$defs/condition"}},
						"postconditions":{"type":"array","minItems":1,"items":{"$ref":"#/$defs/condition"}},
						"idempotency":{"const":"idempotent"},"side_effect":{"const":"browser_state"}
					},
					"required":["action","intent","target","value","preconditions","postconditions","idempotency","side_effect"]
				},
				"wait_for_step":{
					"type":"object","additionalProperties":false,
					"properties":{
						"action":{"const":"wait_for"},"intent":{"type":"string","minLength":1,"maxLength":500},
						"target":{"type":"string","minLength":1},"timeout_ms":{"type":"integer","minimum":1,"maximum":60000},
						"page_state":{"type":["string","null"]},"target_strategy":{"$ref":"#/$defs/target_strategy"},
						"preconditions":{"type":"array","items":{"$ref":"#/$defs/condition"}},
						"postconditions":{"type":"array","items":{"$ref":"#/$defs/condition"}},
						"idempotency":{"const":"idempotent"},"side_effect":{"const":"none"}
					},
					"required":["action","intent","target","preconditions","postconditions","idempotency","side_effect"]
				},
				"assert_text_step":{
					"type":"object","additionalProperties":false,
					"properties":{
						"action":{"const":"assert_text"},"intent":{"type":"string","minLength":1,"maxLength":500},
						"target":{"type":"string","minLength":1},"value":{"type":"string","minLength":1},
						"page_state":{"type":["string","null"]},"target_strategy":{"$ref":"#/$defs/target_strategy"},
						"preconditions":{"type":"array","items":{"$ref":"#/$defs/condition"}},
						"postconditions":{"type":"array","items":{"$ref":"#/$defs/condition"}},
						"idempotency":{"const":"idempotent"},"side_effect":{"const":"none"}
					},
					"required":["action","intent","target","value","preconditions","postconditions","idempotency","side_effect"]
				},
				"assert_url_step":{
					"type":"object","additionalProperties":false,
					"properties":{
						"action":{"const":"assert_url_contains"},"intent":{"type":"string","minLength":1,"maxLength":500},
						"target":{"type":"string","minLength":1},"value":{"type":"string","minLength":1},
						"preconditions":{"type":"array","items":{"$ref":"#/$defs/condition"}},
						"postconditions":{"type":"array","items":{"$ref":"#/$defs/condition"}},
						"idempotency":{"const":"idempotent"},"side_effect":{"const":"none"}
					},
					"required":["action","intent","target","value","preconditions","postconditions","idempotency","side_effect"]
				},
				"capture_text_step":{
					"type":"object","additionalProperties":false,
					"properties":{
						"action":{"const":"capture_text"},"intent":{"type":"string","minLength":1,"maxLength":500},
						"target":{"type":"string","minLength":1},"context_key":{"type":"string","minLength":1},
						"page_state":{"type":["string","null"]},"target_strategy":{"$ref":"#/$defs/target_strategy"},
						"preconditions":{"type":"array","items":{"$ref":"#/$defs/condition"}},
						"postconditions":{"type":"array","items":{"$ref":"#/$defs/condition"}},
						"idempotency":{"const":"idempotent"},"side_effect":{"const":"none"}
					},
					"required":["action","intent","target","context_key","preconditions","postconditions","idempotency","side_effect"]
				},
				"target_strategy":{"type":["string","null"],"enum":["css","xpath","data-testid","element_id","tag",null]},
				"selector_evidence":{
					"type":"object","additionalProperties":false,
					"properties":{"strategy":{"type":"string"},"selector":{"type":"string"},"name":{"type":"string"},"source":{"type":"string"}},
					"required":["selector"]
				},
				"dom":{
					"type":"object","additionalProperties":false,
					"properties":{
						"tag":{"type":"string"},
						"attrs":{"type":"object","patternProperties":{"^.*$":{"type":"string"}},"additionalProperties":false}
					}
				},
				"a11y_node":{
					"type":"object","additionalProperties":false,
					"properties":{
						"node_id":{"type":"string"},"backend_dom_node_id":{},
						"parent_id":{"type":"string"},"role":{"type":"string"},"name":{"type":"string"},
						"page_state":{"type":"string"},"source":{"type":"string"},
						"focusable":{"type":"boolean"},"disabled":{"type":"boolean"},
						"dom":{"$ref":"#/$defs/dom"},
						"verified_selectors":{"type":"array","items":{"$ref":"#/$defs/selector_evidence"}}
					}
				}
			}
		}`),
	}
}

func (t GenerateDSLTool) Execute(ctx context.Context, call Call) (Result, error) {
	content, err := t.client.GenerateDSL(
		ctx,
		call.ActorUserID,
		call.RunID,
		call.ProjectID,
		call.ConversationID,
		call.Arguments,
	)
	if err != nil {
		return Result{}, err
	}
	var generated struct {
		GenerationID int64 `json:"generation_id"`
	}
	if err := json.Unmarshal(content, &generated); err != nil {
		return Result{}, err
	}
	result := Result{Content: content}
	if generated.GenerationID > 0 {
		result.Artifact = &Artifact{
			Type: "dsl_generation",
			ID:   strconv.FormatInt(generated.GenerationID, 10),
		}
	}
	return result, nil
}

File: browser-worker/app/schemas/dsl.py
"""Structured DSL schemas for runnable test cases."""

from __future__ import annotations

import hashlib
import json
from typing import Any, Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


DSL_CANONICAL_VERSION = "dsl.canonical.v1"


class DSLModel(BaseModel):
    model_config = ConfigDict(extra="ignore", str_strip_whitespace=True)


TargetStrategy = Literal[
    "css",
    "xpath",
    "data-testid",
    "element_id",
    "tag",
]
LocatorConfidence = Literal["high", "medium", "low"]


# Strategy name normalization map: AI-generated variant -> canonical name.
_STRATEGY_NORMALIZE: dict[str, str] = {
    "css_selector": "css",
    "data_testid": "data-testid",
    "elementId": "element_id",
    "href": "css",
    "link": "role",
    "button": "role",
    "aria": "role",
    "id": "element_id",
    "name": "tag",
}


class LocatorCandidate(BaseModel):
    """Pre-scored candidate locator strategy for a DSL step."""
    model_config = ConfigDict(extra="ignore", str_strip_whitespace=True)

    strategy: str = Field(description="Locator strategy name (normalized at runtime)")

    @field_validator("strategy", mode="before")
    @classmethod
    def _normalize_strategy(cls, v: str) -> str:
        return _STRATEGY_NORMALIZE.get(v, v)
    selector: str | None = Field(default=None, description="Explicit selector value (for css/xpath/data-testid/etc).")
    semantic_value: str | None = Field(default=None, description="Semantic value (role name, label text, etc).")
    pre_score: float = Field(ge=0.0, le=1.0, description="Generation-time pre-score 0.0-1.0.")
    pre_features: dict | None = Field(default=None, description="Pre-score feature breakdown for debugging.")


class ConditionSpec(BaseModel):
    """Deterministic condition evaluated before or after one action."""
    model_config = ConfigDict(extra="ignore", str_strip_whitespace=True)

    type: Literal[
        "url_contains", "url_changes", "text_visible",
        "text_gone", "element_visible", "element_gone",
        "network_request", "dom_changed", "value_changed",
    ]
    value: str | None = Field(default=None, description="Expected value (URL fragment, text, selector).")
    method: str | None = Field(default=None, description="Expected HTTP method for network_request.")
    status: int | None = Field(default=None, ge=100, le=599, description="Expected HTTP status for network_request.")
    timeout_ms: int = Field(default=3000, ge=100, le=30000)

    @field_validator("method")
    @classmethod
    def normalize_method(cls, value: str | None) -> str | None:
        return value.upper() if value else None

    @model_validator(mode="after")
    def validate_required_value(self) -> "ConditionSpec":
        if self.type == "url_contains" and not self.value:
            raise ValueError("url_contains condition requires a target URL value")
        if self.type == "network_request" and not any((self.value, self.method, self.status)):
            raise ValueError("network_request condition requires URL, method, or status")
        return self


Postcondition = ConditionSpec


class GotoStep(DSLModel):
    action: Literal["goto"]
    value: str = Field(min_length=1, description="Target URL or path.")
    preconditions: list[ConditionSpec] = Field(default_factory=list)
    postconditions: list[ConditionSpec] = Field(default_factory=list)


class ClickStep(DSLModel):
    action: Literal["click"]
    target: str = Field(min_length=1, description="Semantic or explicit locator.")
    page_state: str | None = Field(default=None, description="Page state this step belongs to (S0, S1, …).")
    target_strategy: TargetStrategy | None = Field(default=None, description="Locator strategy hint.")
    locator_confidence: LocatorConfidence | None = Field(
        default=None, description="AI-assessed locator confidence. low triggers VLM pre-verification.",
    )
    candidates: list[LocatorCandidate] = Field(default_factory=list, description="Pre-scored candidate locators.")
    preconditions: list[ConditionSpec] = Field(default_factory=list)
    postconditions: list[ConditionSpec] = Field(default_factory=list, description="Post-action verification conditions.")


class InputStep(DSLModel):
    action: Literal["input"]
    target: str = Field(min_length=1, description="Semantic or explicit locator.")
    value: str = Field(description="Input text.")
    trigger: str | None = Field(default=None, description="Key to press after input (e.g. Enter for quantity/search fields).")
    page_state: str | None = Field(default=None, description="Page state this step belongs to (S0, S1, …).")
    target_strategy: TargetStrategy | None = Field(default=None, description="Locator strategy hint.")
    locator_confidence: LocatorConfidence | None = Field(
        default=None, description="AI-assessed locator confidence. low triggers VLM pre-verification.",
    )
    candidates: list[LocatorCandidate] = Field(default_factory=list, description="Pre-scored candidate locators.")
    preconditions: list[ConditionSpec] = Field(default_factory=list)
    postconditions: list[ConditionSpec] = Field(default_factory=list, description="Post-action verification conditions.")


class WaitForStep(DSLModel):
    action: Literal["wait_for"]
    target: str = Field(min_length=1, description="Target to wait for.")
    timeout_ms: int = Field(default=5000, ge=1, le=60000)
    page_state: str | None = Field(default=None, description="Page state this step belongs to (S0, S1, …).")
    target_strategy: TargetStrategy | None = Field(default=None, description="Locator strategy hint.")
    locator_confidence: LocatorConfidence | None = Field(
        default=None, description="AI-assessed locator confidence. low triggers VLM pre-verification.",
    )
    candidates: list[LocatorCandidate] = Field(default_factory=list, description="Pre-scored candidate locators.")
    preconditions: list[ConditionSpec] = Field(default_factory=list)
    postconditions: list[ConditionSpec] = Field(default_factory=list, description="Post-action verification conditions.")


class AssertTextStep(DSLModel):
    action: Literal["assert_text"]
    target: str = Field(min_length=1, description="Target to assert against.")
    value: str = Field(min_length=1, description="Expected text.")
    page_state: str | None = Field(default=None, description="Page state this step belongs to (S0, S1, …).")
    target_strategy: TargetStrategy | None = Field(default=None, description="Locator strategy hint.")
    locator_confidence: LocatorConfidence | None = Field(
        default=None, description="AI-assessed locator confidence. low triggers VLM pre-verification.",
    )
    candidates: list[LocatorCandidate] = Field(default_factory=list, description="Pre-scored candidate locators.")
    preconditions: list[ConditionSpec] = Field(default_factory=list)
    postconditions: list[ConditionSpec] = Field(default_factory=list, description="Post-action verification conditions.")


class AssertUrlContainsStep(DSLModel):
    action: Literal["assert_url_contains"]
    value: str = Field(min_length=1, description="Expected URL fragment.")
    preconditions: list[ConditionSpec] = Field(default_factory=list)
    postconditions: list[ConditionSpec] = Field(default_factory=list)


class CaptureTextStep(DSLModel):
    action: Literal["capture_text"]
    target: str = Field(min_length=1, description="Element to capture text from.")
    context_key: str = Field(
        min_length=1,
        max_length=100,
        pattern=r"^[A-Za-z_][A-Za-z0-9_]*$",
        description="Runtime variable name to store the captured text.",
    )
    page_state: str | None = Field(default=None, description="Page state this step belongs to (S0, S1, …).")
    target_strategy: TargetStrategy | None = Field(default=None, description="Locator strategy hint.")
    locator_confidence: LocatorConfidence | None = Field(
        default=None, description="AI-assessed locator confidence. low triggers VLM pre-verification.",
    )
    candidates: list[LocatorCandidate] = Field(default_factory=list, description="Pre-scored candidate locators.")
    preconditions: list[ConditionSpec] = Field(default_factory=list)
    postconditions: list[ConditionSpec] = Field(default_factory=list, description="Post-action verification conditions.")


DSLVariableType = Literal["string", "number", "boolean", "object", "array"]
DSLVariableSource = Literal[
    "latest_url",
    "error_message",
    "status",
    "last_step_url",
    "last_step_page_title",
    "last_step_target",
    "last_step_value",
    "last_step_error_message",
]


class DSLCaseInputContract(DSLModel):
    name: str = Field(min_length=1, max_length=100)
    context_key: str = Field(min_length=1, max_length=100, pattern=r"^[A-Za-z_][A-Za-z0-9_]*$")
    value_type: DSLVariableType
    required: bool = True
    description: str | None = Field(default=None, max_length=500)
    value: str | None = Field(default=None, description="Default value for this variable.")


class DSLCaseOutputContract(DSLModel):
    name: str = Field(min_length=1, max_length=100)
    context_key: str = Field(min_length=1, max_length=100, pattern=r"^[A-Za-z_][A-Za-z0-9_]*$")
    value_type: DSLVariableType
    source: DSLVariableSource | None = None
    description: str | None = Field(default=None, max_length=500)


DSLStep = Annotated[
    GotoStep
    | ClickStep
    | InputStep
    | WaitForStep
    | AssertTextStep
    | AssertUrlContainsStep
    | CaptureTextStep,
    Field(discriminator="action"),
]


class DSLCase(DSLModel):
    name: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=1000)
    base_url: str | None = Field(default=None, min_length=1, max_length=500)
    input_contract: list[DSLCaseInputContract] = Field(default_factory=list)
    output_contract: list[DSLCaseOutputContract] = Field(default_factory=list)
    steps: list[DSLStep] = Field(min_length=1)

    def model_dump(self, *args: Any, **kwargs: Any) -> dict[str, Any]:
        payload = super().model_dump(*args, **kwargs)
        for index, step in enumerate(self.steps):
            step_payload = payload["steps"][index]
            for field_name in ("preconditions", "postconditions"):
                if field_name not in step.model_fields_set:
                    step_payload.pop(field_name, None)
                    continue
                conditions = getattr(step, field_name)
                for condition_index, condition in enumerate(conditions):
                    for condition_field in ("method", "status"):
                        if condition_field not in condition.model_fields_set:
                            step_payload[field_name][condition_index].pop(
                                condition_field,
                                None,
                            )
        return payload


def load_canonical_dsl(
    canonical_json: str,
    expected_sha256: str,
    canonical_version: str,
) -> tuple[DSLCase, dict]:
    """Verify Go-owned canonical bytes and reject semantic normalization drift."""
    if canonical_version != DSL_CANONICAL_VERSION:
        raise ValueError(f"Unsupported DSL canonical version: {canonical_version}.")
    actual_sha256 = hashlib.sha256(canonical_json.encode("utf-8")).hexdigest()
    if actual_sha256 != expected_sha256:
        raise ValueError("Canonical DSL SHA-256 does not match the approved generation.")
    payload = json.loads(canonical_json)
    if not isinstance(payload, dict):
        raise ValueError("Canonical DSL must be a JSON object.")
    case = DSLCase.model_validate(payload)
    materialized = case.model_dump(mode="json")
    for index, source_step in enumerate(payload.get("steps", [])):
        for compatibility_field in ("preconditions", "postconditions"):
            if compatibility_field not in source_step:
                materialized["steps"][index].pop(compatibility_field, None)
                continue
            for condition_index, source_condition in enumerate(
                source_step[compatibility_field]
            ):
                for condition_field in ("method", "status"):
                    if condition_field not in source_condition:
                        materialized["steps"][index][compatibility_field][
                            condition_index
                        ].pop(condition_field, None)
    if materialized != payload:
        raise ValueError("Canonical DSL is not fully materialized or violates the worker schema.")
    return case, payload
